src/tactic/norm_num.lean

```diff
@@ -1384,33 +1384,86 @@ meta def eval_prime : expr → tactic (expr × expr)
 | _ := failed
 
 /-- This version of `derive` does not fail when the input is already a numeral -/
-meta def derive' (e : expr) : tactic (expr × expr) :=
+meta def derive.step (e : expr) : tactic (expr × expr) :=
 eval_field e <|> eval_nat_int_ext e <|>
 eval_pow e <|> eval_ineq e <|> eval_prime e
 
-meta def derive : expr → tactic (expr × expr) | e :=
+/-- An attribute for adding additional extensions to `norm_num`. To use this attribute, put
+`@[norm_num]` on a tactic of type `expr → tactic (expr × expr)`; the tactic will be called on
+subterms by `norm_num`, and it is responsible for identifying that the expression is a numerical
+function applied to numerals, for example `nat.fib 17`, and should return the reduced numerical
+expression (which must be in `norm_num`-normal form: a natural or rational numeral, i.e. `37`,
+`12 / 7` or `-(2 / 3)`, although this can be an expression in any type), and the proof that the
+original expression is equal to the rewritten expression. -/
```

Comment:
Could you clarify the failure behavior of the function? Each function recognizes a certain pattern, and should fail when the input doesn't match this pattern, right? There's no way to report an exception differently from "this doesn't match the pattern"?

Reply:
In fact, what kind of `expr` is this being called on, `Prop`s or terms? Both? I see in your test case it's used on a term, which is what I assumed at first, but if this is being used to handle the `prime 7` extension, maybe not?

Reply:
There is no support for other kinds of failure. If a norm_num extension tactic fails, it moves on to the next one, so it is important for performance that they fail fast. So generally you first check that the term has the right form, then you construct the proof, and the proof construction stage should not fail (it can, but this will just be interpreted as the simplification routine not being applicable).

It gets called on props and terms. For props like `prime 7`, you return a proof of `prime 7 = true` just like everything else.

Reply:
Right, could you add this info to the docs?